Compiling Himmelblau on Ubuntu 24.04

I'm following the instructions from the homepage to install Ubuntu packages. The only difference is that I had to use `libclang-dev` instead of `libclang-13-dev`. On the fresh Ubuntu 24.04 LTS Desktop VM, I've installed the Ubuntu packages:

```bash
sudo apt-get install libpam0g-dev libudev-dev libssl-dev pkg-config tpm-udev libtss2-dev libcap-dev libtalloc-dev libtevent-dev libldb-dev libdhash-dev libkrb5-dev libpcre2-dev libclang-dev autoconf gettext
```

I've used rustup to install Rust toolchain.

The `make` process fails as follows:

```makefile
error: failed to run custom build command for `idmap v0.5.0 (/home/aleksandar-miljkovic/AD/himmelblau/src/idmap)`

Caused by:
  process didn't exit successfully: `/home/user/AD/himmelblau/target/debug/build/idmap-afd29ee2208ec77e/build-script-build` (exit status: 101)
  --- stderr
  sssd/src/external/libpcre.m4:20: warning: AC_LIB_PREPARE_PREFIX is m4_require'd but not m4_defun'd
  sssd/src/build_macros.m4:42: SSS_AC_EXPAND_LIB_DIR is expanded from...
  sssd/src/external/libpcre.m4:20: the top level
  configure:4374: error: possibly undefined macro: AC_LIB_PREPARE_PREFIX
        If this token and others are legitimate, please use m4_pattern_allow.
        See the Autoconf documentation.
  configure:4377: error: possibly undefined macro: AC_LIB_WITH_FINAL_PREFIX
  autoreconf: error: /usr/bin/autoconf failed with exit status: 1
  thread 'main' panicked at src/idmap/build.rs:14:9:
  Failed to configure sss_idmap
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
make: *** [Makefile:3: all] Error 101
```

Any suggestions on fixing this? I would like to build Ubuntu packages and test them on 24.04, 22.04 and possibly 20.04.

Thank you in advance!

Same here on current Debian testing. It seems, libhimmelblau 0.2.7+ has gone incompatible with himmelblau by renaming itself from msal to himmelblau and changing at least some of its API. Pinning it to =0.2.6 in Cargo.toml is a workaround. Upping it to 0.2.7 and changing msal:: to himmelblau:: in `src/common/src/idprovider/himmelblau.rs` gets me as far as:
``` make
   Compiling himmelblau_unix_common v0.5.0 (/home/m/himmelblau/src/common)                                                                                                                                                                                                        
error[E0308]: mismatched types                                                                                                                                                                                                                                                    
    --> src/common/src/idprovider/himmelblau.rs:1426:28                                                                                                                                                                                                                           
     |                                                                                                                                                                                                                                                                            
1426 |             .enroll_device(token, attrs, tpm, machine_key)                                                                                                                                                                                                                 
     |              ------------- ^^^^^ expected `&str`, found `&UserToken`                                                                                                                                                                                                       
     |              |                                                                                                                                                                                                                                                             
     |              arguments to this method are incorrect                                                                                                                                                                                                                        
     |                                                                                                                                                                                                                                                                            
     = note: expected reference `&str`                                                                                                                                                                                                                                            
                found reference `&himmelblau::UserToken`                                                                                                                                                                                                                          
note: method defined here                                                                                                                                                                                                                                                         
    --> /home/m/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libhimmelblau-0.2.9/src/auth.rs:1977:18                                                                                                                                                                      
     |                                                                                                                                                                                                                                                                            
1977 |     pub async fn enroll_device(                                                                                                                                                                                                                                            
     |                  ^^^^^^^^^^^^^                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                  
For more information about this error, try `rustc --explain E0308`.             
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -36,7 +36,7 @@ tracing-subscriber = "^0.3.17"
 tracing = "^0.1.37"
 himmelblau_unix_common = { path = "src/common" }
 kanidm_unix_common = { path = "src/glue" }
-libhimmelblau = { version = "0.2.6" }
+libhimmelblau = { version = "0.2.9" }
 clap = { version = "^4.5", features = ["derive", "env"] }
 clap_complete = "^4.4.1"
 reqwest = { version = "^0.12.2", features = ["json"] }

--- a/src/common/src/idprovider/himmelblau.rs
+++ b/src/common/src/idprovider/himmelblau.rs
@@ -10,16 +10,16 @@ use crate::idprovider::interface::tpm;
 use crate::unix_proto::{DeviceAuthorizationResponse, PamAuthRequest};
 use anyhow::{anyhow, Result};
 use async_trait::async_trait;
-use idmap::SssIdmap;
-use kanidm_hsm_crypto::{LoadableIdentityKey, LoadableMsOapxbcRsaKey, PinValue, SealedData, Tpm};
-use msal::auth::{
+use himmelblau::auth::{
     BrokerClientApplication, ClientInfo,
     DeviceAuthorizationResponse as msal_DeviceAuthorizationResponse, IdToken, MFAAuthContinue,
     UserToken as UnixUserToken,
 };
-use msal::discovery::EnrollAttrs;
-use msal::error::{ErrorResponse, MsalError, AUTH_PENDING, DEVICE_AUTH_FAIL, REQUIRES_MFA};
-use msal::graph::{DirectoryObject, Graph};
+use himmelblau::discovery::EnrollAttrs;
+use himmelblau::error::{ErrorResponse, MsalError, AUTH_PENDING, DEVICE_AUTH_FAIL, REQUIRES_MFA};
+use himmelblau::graph::{DirectoryObject, Graph};
+use idmap::SssIdmap;
+use kanidm_hsm_crypto::{LoadableIdentityKey, LoadableMsOapxbcRsaKey, PinValue, SealedData, Tpm};
 use reqwest;
 use std::collections::HashMap;
 use std::sync::Arc;
@@ -1423,7 +1423,7 @@ impl HimmelblauProvider {
             .client
             .write()
             .await
-            .enroll_device(token, attrs, tpm, machine_key)
+            .enroll_device(&token.refresh_token, attrs, tpm, machine_key)
             .await
         {
             Ok((new_loadable_transport_key, new_loadable_cert_key, device_id)) => {

--- a/src/nss/src/lib.rs
+++ b/src/nss/src/lib.rs
@@ -14,6 +14,5 @@
 #[macro_use]
 extern crate libnss;
 #[cfg(target_family = "unix")]
-
 #[cfg(target_family = "unix")]
 mod implementation;

